### ISSUE TITLE
Core dump while processing a 33GB dataset

### ISSUE THREAD

**Issue by david-cliqz** [cliqz-oss/keyvi#40]
This is probably the biggest dataset we have tried yet. The error happens at around 88% (`Progress: 223600000/253842680` according to Jenkins). The backtrace is (excluding the Python fluff):
```
#0  __memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:167
#1  0x00007efd82de9ceb in memcpy (__len=4, __src=0x7fff198ea2d0, __dest=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/string3.h:51
#2  WriteRawValue (length=4, buffer=0x7fff198ea2d0, offset=<optimized out>, this=0x2dce340) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/internal/sparse_array_persistence.h:160
#3  keyvi::dictionary::fsa::internal::SparseArrayBuilder<keyvi::dictionary::fsa::internal::SparseArrayPersistence<unsigned short> >::WriteTransition (this=0x37f59a0, offset=0, 
    transitionId=<optimized out>, transitionPointer=<optimized out>) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/internal/sparse_array_builder.h:408
#4  0x00007efd82e01849 in WriteState (unpacked_state=..., offset=4294967180, this=0x37f59a0) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/internal/sparse_array_builder.h:246
#5  keyvi::dictionary::fsa::internal::SparseArrayBuilder<keyvi::dictionary::fsa::internal::SparseArrayPersistence<unsigned short> >::PersistState (this=<optimized out>, unpacked_state=...)
    at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/internal/sparse_array_builder.h:86
#6  0x00007efd82e2517e in ConsumeStack (end=<optimized out>, this=<optimized out>) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/generator.h:371
#7  Add<std::basic_string<char> > (value=..., input_key=..., this=0x3372fc8) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/generator.h:218
#8  keyvi::dictionary::DictionaryCompiler<keyvi::dictionary::fsa::internal::SparseArrayPersistence<unsigned short>, keyvi::dictionary::fsa::internal::JsonValueStore>::Compile(std::function<void (unsigned long, unsigned long, void*)>, void*) (this=0x3372e50, progress_callback=..., user_data=user_data@entry=0x7efd704a6b50)
    at /raid/keyvi/keyvi/src/cpp/dictionary/dictionary_compiler.h:138
#9  0x00007efd82ddbda4 in __pyx_pf_7pykeyvi_22JsonDictionaryCompiler_18Compile (__pyx_v_args=0x7efd7044e050, __pyx_v_self=0x7efd70869d90) at src/pykeyvi.cpp:3946
#10 __pyx_pw_7pykeyvi_22JsonDictionaryCompiler_19Compile (__pyx_v_self=0x7efd70869d90, __pyx_args=0x7efd7044e050, __pyx_kwds=<optimized out>) at src/pykeyvi.cpp:3835
```

```
# dpkg -l | grep keyvi
ii  keyvi              0.1.3         amd64        key value index
ii  python-keyvi       0.1.4         amd64        python bindings for keyvi
```
Could this be the problem?

**Comment by david-cliqz** [comment]
OK, a better look at the backtrace:
```
#3  keyvi::dictionary::fsa::internal::SparseArrayBuilder<keyvi::dictionary::fsa::internal::SparseArrayPersistence<unsigned short> >::WriteTransition (this=0x37f59a0, offset=0, 
    transitionId=<optimized out>, transitionPointer=<optimized out>) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/internal/sparse_array_builder.h:408
        transitionPointer_low = 2
        vshort_pointer = {65521, 16383, 0, 0, 23432, 895, 0, 0}
        start_position = 1000
        overflow_bucket = <optimized out>
        difference = <optimized out>
        pt_to_overflow_bucket = 32768
        overflow_code = <optimized out>
        transitionPointer_high = <optimized out>
        vshort_size = <optimized out>
#4  0x00007efd82e01849 in WriteState (unpacked_state=..., offset=4294967180, this=0x37f59a0) at /raid/keyvi/keyvi/src/cpp/dictionary/fsa/internal/sparse_array_builder.h:246
        e = {label = <optimized out>, value = 4294967178}
        i = <optimized out>
        len = <optimized out>
        weight = 0
```
revealed to us that we simply ran out of node ids. We should experiment with 64 bit keys / 32 compressed (or recursively compressed?)

**Comment by hendrik-cliqz** [comment]
The keyvi persistence layer should be capable for larger sizes as it uses variable length integers. 

This issue is in the compiler: First the big minimization hashtable: 1 bucket has indeed a limitation of 32bits for the transition offset. Because doubling this pointer to 64bit would increase the hashtable significantly, which means increasing memory consumption significantly. Therefore I will try to implement it some smartness and only use 64bit pointers if needed.

**Comment by hendrik-cliqz** [comment]
fixed with 0.1.9
